packages/tocco-util/src/form/formDefinition.spec.js

```diff
@@ -0,0 +1,147 @@
+      beforeEach(() => {
+        fetchMock.reset()
+        fetchMock.restore()
+      })
+
+      describe('getFieldDefinitions', () => {
+        it('return an array of fields', () => {
+          const fields = formDefinition.getFieldDefinitions(testFormDefinition)
+          expect(fields).to.eql([testField1, testField2, testDisplay, testField3])
+        })
+
+        it('should return an empty array in case of no fields', () => {
+          const fields = formDefinition.getFieldDefinitions({id: 'fromX', children: []})
+          expect(fields).to.eql([])
+        })
+      })
+
+      describe('getFieldNames', () => {
+        it('should return a list of all field names', () => {
+          const fields = formDefinition.getFieldDefinitions(testFormDefinition)
+          const fieldNames = formDefinition.getFieldNames(fields)
+          expect(fieldNames).to.eql([testField1.id, testField2.id, testDisplay.id, testField3.id])
+        })
+      })
+
+      describe('getDefaultValues', () => {
+        it('should return all default values', () => {
+          const fields = formDefinition.getFieldDefinitions(testFormDefinition)
+          const fieldNames = formDefinition.getDefaultValues(fields)
+          expect(fieldNames).to.eql({
+            [testField2.id]: testField2.defaultValue,
+            [testField3.id]: testField3.defaultValue
+          })
+        })
+      })
+
+      describe('fetchForm', () => {
+        it('should call request saga and transform response', () => {
+          const gen = formDefinition.fetchForm('User_search')
+
+          expect(gen.next().value).to.eql(call(requestSaga, 'forms/User_search'))
+
+          const resp = {
+            body: {
+              form: {
+                children: [{
+                  name: 'firstname',
+                  type: 'ch.tocco.nice2.model.form.components.simple.TextArea'
+                }]
+              }
+            }
+          }
+
+          expect(gen.next(resp).value).to.eql(call(formDefinition.defaultFormTransformer, resp.body))
+
+          const next = gen.next(resp.body.form)
+
+          expect(next.value).to.eql(resp.body.form)
+          expect(next.done).to.be.true
+        })
+      })
+      describe('defaultFormTransformer', () => {
+        it('should return form without table overhead', () => {
+          const form = {
+            id: 'fromX',
+            children: [
+              {}
+            ]
+          }
+          const result = formDefinition.defaultFormTransformer({form})
+          expect(result).to.eql(form)
+        })
+      })
+
+      describe('getFieldId', () => {
+        it('should return a string containing formName and fieldName', () => {
+          const formName = 'User_detail'
+          const fieldName = 'firstname'
+          const result = formDefinition.getFieldId(formName, fieldName)
+          expect(result).to.have.string(formName)
+          expect(result).to.have.string(fieldName)
```

Comment:
Are we using the generated ID ("input-{formName}-{fieldName}") just for simple identification - or does it matter what it looks like exactly?
For example, if the ID will be used for styling purposes as well (CSS selectors) it would in my opinion make sense to check the generated ID explicitly against the actual format ("input-{formName}-{fieldName}").
(now the test implies that it really doesn't matter how the ID looks like, as long as the form name and the field name are in there - and that we can change the actual format without having to fear any side effects)

Reply:
at the moment this should generate an unique id for each field on the form. We're using it to set the focus on these fields if there is a validation error.
Since we rewrite the application to styled components i think css selector compatibility can be ignored.